src/main/scala/no/ndla/imageapi/service/ImageConverter.scala

```diff
@@ -24,48 +41,89 @@ trait ImageConverter {
     }
 
     def resize(originalImage: ImageStream, mode: Mode, targetSize: Int): Try[ImageStream] = {
-      val sourceImage = ImageIO.read(originalImage.stream)
-      val maxSize = mode match {
-        case Mode.FIT_TO_WIDTH => sourceImage.getWidth
-        case Mode.FIT_TO_HEIGHT => sourceImage.getHeight
-        case _ => max(sourceImage.getWidth, sourceImage.getHeight)
-      }
-      Try(Scalr.resize(sourceImage, mode, min(targetSize, maxSize))).map(resized => toImageSteam(resized, originalImage))
+      val MaxTargetSize = 2000
+      val sourceImage = originalImage.sourceImage
+      Try(Scalr.resize(sourceImage, mode, min(targetSize, MaxTargetSize))).map(resized => toImageSteam(resized, originalImage))
     }
 
     def resizeWidth(originalImage: ImageStream, size: Int): Try[ImageStream] = resize(originalImage, Mode.FIT_TO_WIDTH, size)
 
     def resizeHeight(originalImage: ImageStream, size: Int): Try[ImageStream] = resize(originalImage, Mode.FIT_TO_HEIGHT, size)
 
-    def crop(originalImage: ImageStream, start: Point, end: Point): Try[ImageStream] = {
-      val sourceImage = ImageIO.read(originalImage.stream)
-      val (topLeft, bottomRight) = transformCoordinates(start, end)
+    private def crop(image: ImageStream, sourceImage: BufferedImage, topLeft: PixelPoint, bottomRight: PixelPoint): Try[ImageStream] = {
       val (width, height) = getWidthHeight(topLeft, bottomRight, sourceImage)
 
       Try(Scalr.crop(sourceImage, topLeft.x, topLeft.y, width, height))
-        .map(cropped => toImageSteam(cropped, originalImage))
+        .map(cropped => toImageSteam(cropped, image))
+    }
+
+    def crop(originalImage: ImageStream, start: PercentPoint, end: PercentPoint): Try[ImageStream] = {
+      val sourceImage = originalImage.sourceImage
+      val (topLeft, bottomRight) = transformCoordinates(sourceImage, start, end)
+      crop(originalImage, sourceImage, topLeft, bottomRight)
+    }
+
+    private def getStartEnd(size: Int, targetSize: Int, focal: Int): (Int, Int) = {
+      val ts = min(targetSize, size) / 2
+      val (start, end) = (focal - ts, focal + ts)
+      val (startRem, endRem) = (abs(min(start, 0)), max(end - size, 0))
+
+      (max(start - endRem, 0), min(end + startRem, size))
+    }
+
+    def dynamicCrop(image: ImageStream, percentFocalPoint: PercentPoint, targetWidth: Int, targetHeight: Int): Try[ImageStream] = {
+      val sourceImage = image.sourceImage
+      val focalPoint = toPixelPoint(percentFocalPoint, sourceImage)
+      val (imageWidth, imageHeight) = (sourceImage.getWidth, sourceImage.getHeight)
+
+      val (startY, endY) = getStartEnd(imageHeight, targetHeight, focalPoint.y)
+      val (startX, endX) = getStartEnd(imageWidth, targetWidth, focalPoint.x)
+
+      crop(image, sourceImage, PixelPoint(startX, startY), PixelPoint(endX, endY))
+    }
+
+    def dynamicCropWidth(image: ImageStream, percentFocalPoint: PercentPoint, targetWidth: Int): Try[ImageStream] = {
```

Comment:
Mulig intensjonen var fornuftig i å skille `dynamicCropWidth` og `dynamicCropHeight` ut i egne funksjoner. Men jeg brukte lang tid på å skjønne hvor like de faktisk er. Tror kanskje det hadde vært lettere å lese logikken dersom alle tre `dynamicCrop*`-funksjonene var slått sammen i én, med ordentlige "grener" for om man forholder seg til fokuspunkt, bredde eller høyde når man velger verdier å gå ut ifra...